readme.md
# canvas 思维导图 




## 值得注意的该导图支持多个根节点，也就是多个树

### 实现功能：
#### 1.节点可展开收起
#### 2.图可以拖动平移、鼠标缩放

![image](https://github.com/LucyHeres/canvas-mind/blob/master/canva-%E6%80%9D%E7%BB%B4%E5%AF%BC%E5%9B%BE.jpg)
#### 实现方法：
```js
// 树图属性数据：
var options = {
  container: "#canvas-container", //容器
  keyNodeWidth: 320, //节点宽度
  keyNodeHeight: 100, //节点高度
  hubRadius: 10, //枢纽小圆点半径
  lineColor: "#e3e4e5", //节点间连线 线条颜色
};

// 节点数据：（如下：两个根节点）
var nodeArray = [
  {
    id: "1",
    isRoot: true,
    expandedLeft: false,
    expandedRight: false,
    childrenCountLeft: 2,
    childrenCountRight: 1,
    children: [
      {
        id: "1.1",
        expanded: false,
        direction: -1,
        childrenCount: 2,
        children: [
          {
            id: "1.1.1",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
          {
            id: "1.1.2",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
        ],
      },
      {
        id: "1.2",
        expanded: false,
        direction: -1,
        childrenCount: 2,
        children: [
          {
            id: "1.2.1",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
          {
            id: "1.2.2",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
        ],
      },
      {
        id: "1.3",
        expanded: false,
        direction: 1,
        childrenCount: 2,
        children: [
          {
            id: "1.3.1",
            expanded: false,
            direction: 1,
            childrenCount: 0,
            children: [],
          },
          {
            id: "1.3.2",
            expanded: false,
            direction: 1,
            childrenCount: 0,
            children: [],
          },
        ],
      },
    ],
  },
  {
    id: "2",
    isRoot: true,
    expandedLeft: false,
    expandedRight: false,
    childrenCountLeft: 2,
    childrenCountRight: 1,
    children: [
      {
        id: "2.1",
        expanded: false,
        direction: -1,
        childrenCount: 2,
        children: [
          {
            id: "2.1.1",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
          {
            id: "2.1.2",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
        ],
      },
      {
        id: "2.2",
        expanded: false,
        direction: -1,
        childrenCount: 2,
        children: [
          {
            id: "2.2.1",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
          {
            id: "2.2.2",
            expanded: false,
            direction: -1,
            childrenCount: 0,
            children: [],
          },
        ],
      },
      {
        id: "2.3",
        expanded: false,
        direction: 1,
        childrenCount: 2,
        children: [
          {
            id: "2.3.1",
            expanded: false,
            direction: 1,
            childrenCount: 0,
            children: [],
          },
          {
            id: "2.3.2",
            expanded: false,
            direction: 1,
            childrenCount: 0,
            children: [],
          },
        ],
      },
    ],
  },
];
// 此数据只实现了树图结构的显示，节点为空白矩形，如果需要节点中展示文字或图形，则需要在每个节点中传入图形文字数据对象。
// 如下：
var text = [
  {
    value: "文字一",
    top: 20,
    left: 54,
    width: 60,
    height: 16,
    fontsize: 16,
    lineHeight: 16,
    color: "#666666",
  },
  // ...
];
var shape = [
  {
    type: "rect",
    background: "#e4393c",
    left: 0,
    top: 0,
    width: 3,
    height: 100,
  },
  // ...
];
// 讲以上两个对象传入每个节点中，即可实现节点中图形文字的显示。具体可参考 代码文件 parse.js

var qjMind = new qjMind(options, {
  keyNodeClick: function (node) {
    console.log("点击了okr节点", node.id, node.x, node.y);
    //
  },
  hubNodeClick: function (node, dir) {
    console.log("点击了hub节点", node);

    //TODO: 改变数据，改变节点展开收起属性

    // 改变后重新渲染
    qjMind.render(nodeArray);
  },
});
qjMind.render(nodeArray);
```


CHANGELOG.md
## [1.6.27-0](https://github.com/LucyHeres/canvas-mind/compare/v1.6.26...v1.6.27-0) (2022-08-09)



## [1.6.26](https://github.com/LucyHeres/canvas-mind/compare/v1.6.19...v1.6.26) (2022-08-09)

### Issues
[#33](https://github.com/LucyHeres/canvas-mind/issues/33),[#99](https://github.com/LucyHeres/canvas-mind/issues/99)


### Features

* hhhfd ([a9fd560](https://github.com/LucyHeres/canvas-mind/commit/a9fd560004ec8ed04d04b595fe8660fdee54b5ca))
* hhhhhhh ([4e0c4fd](https://github.com/LucyHeres/canvas-mind/commit/4e0c4fd604c63573393d0d23ccb47349b2c4c01b)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)



## [1.6.25](https://github.com/LucyHeres/canvas-mind/compare/v1.6.19...v1.6.25) (2022-08-09)

### Issues
[#33](https://github.com/LucyHeres/canvas-mind/issues/33),[#99](https://github.com/LucyHeres/canvas-mind/issues/99)


### Features

* hhhhhhh ([4e0c4fd](https://github.com/LucyHeres/canvas-mind/commit/4e0c4fd604c63573393d0d23ccb47349b2c4c01b)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)



## [1.6.24](https://github.com/LucyHeres/canvas-mind/compare/v1.6.19...v1.6.24) (2022-08-09)

### Issues
[#33](https://github.com/LucyHeres/canvas-mind/issues/33),[#99](https://github.com/LucyHeres/canvas-mind/issues/99)


### Features

* hhhhhhh ([4e0c4fd](https://github.com/LucyHeres/canvas-mind/commit/4e0c4fd604c63573393d0d23ccb47349b2c4c01b)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)



## [1.6.23](https://github.com/LucyHeres/canvas-mind/compare/v1.6.19...v1.6.23) (2022-08-09)


### Features

* hhhhhhh ([4e0c4fd](https://github.com/LucyHeres/canvas-mind/commit/4e0c4fd604c63573393d0d23ccb47349b2c4c01b)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)



## [1.6.22](https://github.com/LucyHeres/canvas-mind/compare/v1.6.19...v1.6.22) (2022-08-09)


### Features

* hhhhhhh ([4e0c4fd](https://github.com/LucyHeres/canvas-mind/commit/4e0c4fd604c63573393d0d23ccb47349b2c4c01b)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)



## [1.6.21](https://github.com/LucyHeres/canvas-mind/compare/v1.6.19...v1.6.21) (2022-08-09)


### Features

* hhhhhhh ([4e0c4fd](https://github.com/LucyHeres/canvas-mind/commit/4e0c4fd604c63573393d0d23ccb47349b2c4c01b)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)



## [1.6.20](https://github.com/LucyHeres/canvas-mind/compare/v1.6.19...v1.6.20) (2022-08-09)


### Features

* hhhhhhh ([4e0c4fd](https://github.com/LucyHeres/canvas-mind/commit/4e0c4fd604c63573393d0d23ccb47349b2c4c01b)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)



## [1.6.19](https://github.com/LucyHeres/canvas-mind/compare/v1.6.14...v1.6.19) (2022-08-09)


### Features

* hhhhhhh ([0e85f9c](https://github.com/LucyHeres/canvas-mind/commit/0e85f9cdc292bb7bf91f30305afcadf004bae3bb)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33) [#99](https://github.com/LucyHeres/canvas-mind/issues/99)
* hhhhhhh ([8ee7f96](https://github.com/LucyHeres/canvas-mind/commit/8ee7f961f1f2243ade60235227deab9032c3f12c)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)
* hhhhhhh ([db34729](https://github.com/LucyHeres/canvas-mind/commit/db347292bb0416e2e36baa8fd15b551190388c01)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)



## [1.6.18](https://github.com/LucyHeres/canvas-mind/compare/v1.6.14...v1.6.18) (2022-08-09)


### Features

* hhhhhhh ([8ee7f96](https://github.com/LucyHeres/canvas-mind/commit/8ee7f961f1f2243ade60235227deab9032c3f12c)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)
* hhhhhhh ([db34729](https://github.com/LucyHeres/canvas-mind/commit/db347292bb0416e2e36baa8fd15b551190388c01)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)



## [1.6.17](https://github.com/LucyHeres/canvas-mind/compare/v1.6.14...v1.6.17) (2022-08-09)


### Features

* hhhhhhh ([8ee7f96](https://github.com/LucyHeres/canvas-mind/commit/8ee7f961f1f2243ade60235227deab9032c3f12c)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)
* hhhhhhh ([db34729](https://github.com/LucyHeres/canvas-mind/commit/db347292bb0416e2e36baa8fd15b551190388c01)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)



## [1.6.16](https://github.com/LucyHeres/canvas-mind/compare/v1.6.14...v1.6.16) (2022-08-09)


### Features

* hhhhhhh ([8ee7f96](https://github.com/LucyHeres/canvas-mind/commit/8ee7f961f1f2243ade60235227deab9032c3f12c)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)
* hhhhhhh ([db34729](https://github.com/LucyHeres/canvas-mind/commit/db347292bb0416e2e36baa8fd15b551190388c01)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)



## [1.6.15](https://github.com/LucyHeres/canvas-mind/compare/v1.6.14...v1.6.15) (2022-08-09)


### Features

* hhhhhhh ([db34729](https://github.com/LucyHeres/canvas-mind/commit/db347292bb0416e2e36baa8fd15b551190388c01)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)



## [1.6.14](https://github.com/LucyHeres/canvas-mind/compare/v1.6.13...v1.6.14) (2022-08-09)


### Features

* hhhhhhh ([319798f](https://github.com/LucyHeres/canvas-mind/commit/319798f44a6430c866a6db235f734c73e401c69d)), closes [#33](https://github.com/LucyHeres/canvas-mind/issues/33)



## [1.6.13](https://github.com/LucyHeres/canvas-mind/compare/v1.6.12...v1.6.13) (2022-08-09)


### Features

* Merge branch 85-xin-feature-score into master'+ ([80abe51](https://github.com/LucyHeres/canvas-mind/commit/80abe5140a929b58d66f3d63bcdb6770818ed79b))




parse.js
var nodeObj = {
  employeeId: "uuid",
  employeeName: "张三生三世",
  departmentName: "策划范德萨范德萨加啊部",
  dangerType: 2,
  schedule: "300.00",
  objectiveDesc: "1",
  superCount: 2,
  childCount: 17,
  superOkr: [
    {
      employeeId: "uuid",
      employeeName: "张三生三世",
      departmentName: "策划范德萨范德萨加啊部",
      dangerType: 2,
      schedule: "300.00",
      objectiveDesc: "1.1",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.1",
      superCount: 0,
      childCount: 0,
    },
  ],
  childOkr: [
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
    {
      employeeId: "uuid",
      employeeName: "张",
      departmentName: "策划部",
      dangerType: 2,
      schedule: "30",
      objectiveDesc: "1.2",
      superCount: 0,
      childCount: 0,
    },
  ],
};
var array = [];
// 大数据压测
for (var i = 1; i <= 10; i++) {
  var newObj = JSON.parse(JSON.stringify(nodeObj));
  newObj.id = i;
  newObj.objectiveDesc = "" + i;
  newObj.superOkr.forEach((item,j)=>{
    item.id = "" + i + "."+j;
    item.objectiveDesc = "" + i + "."+j;
  })
  newObj.childOkr.forEach((item,j)=>{
    item.id = "" + i + "."+j;
    item.objectiveDesc = "" + i + "."+j;
  })
  array.push(newObj);
}

const getScheduleInfo = (obj) => {
  let info = {
    0: [`         ${obj.schedule}%`, "#d8d8d8"],
    1: [`正常   ${obj.schedule}%`, "#16A0FF"],
    2: [`有风险 ${obj.schedule}%`, "#FFAE11"],
    3: [`已延期 ${obj.schedule}%`, "#F73838"],
  };
  return info[obj.dangerType];
};
const initRoot = (array) => {
  for (let i = 0; i < array.length; i++) {
    array[i].isRoot = true;
  }
};

const newid = () => {
  return (
    new Date().getTime().toString(16) + Math.random().toString(16).substr(2)
  ).substr(2, 16);
};

/**
 * 格式化接口请求到的目标地图数据
 * @param array
 * @param [dir]  左边的节点：-1 | 右边的节点：1
 */
const parseOkrJSON = (array, dir) => {
  let newArr = [];
  for (let i = 0; i < array.length; i++) {
    let obj = array[i];
    let newObj = {};
    newObj.id = obj.id;
    newObj.nid = newObj.nid || newid();
    if (!("canRead" in obj)) {
      obj.canRead = 1;
    }
    newObj.canRead = obj.canRead;
    if (obj.isRoot) newObj.isRoot = obj.isRoot;
    newObj.text = getTextObj(obj);
    newObj.shape = getShapeObj(obj);
    if (obj.isRoot) {
      newObj.expandedLeft = false;
      newObj.expandedRight = false;
      newObj.childrenCountLeft = obj.superCount;
      newObj.childrenCountRight = obj.childCount;
      newObj.children = [];
      if (obj.superOkr)
        newObj.children = newObj.children.concat(
          parseOkrJSON(obj.superOkr, -1)
        );
      if (obj.childOkr)
        newObj.children = newObj.children.concat(parseOkrJSON(obj.childOkr, 1));
    } else {
      newObj.expanded = false;
      newObj.direction = dir;
      newObj.children = [];
      if (dir === -1 && obj.superCount) {
        newObj.childrenCount = obj.superCount;
        if (obj.superOkr)
          newObj.children = newObj.children.concat(
            parseOkrJSON(obj.superOkr, -1)
          );
      }
      if (dir === 1 && obj.childCount) {
        newObj.childrenCount = obj.childCount;
        if (obj.childOkr)
          newObj.children = newObj.children.concat(
            parseOkrJSON(obj.childOkr, 1)
          );
      }
    }
    newArr.push(newObj);
  }
  return newArr;
};
/**
 * 格式化单次请求到的数组
 * @param array
 * @param dir  左边的节点：-1 | 右边的节点：1
 */
const parseSingleOkrJSON = (array, dir) => {
  let newArr = [];
  for (let i = 0; i < array.length; i++) {
    let newObj = {};
    let obj = array[i];
    newObj.id = obj.id;
    newObj.nid = newObj.nid || newid();
    if (!("canRead" in obj)) {
      obj.canRead = 1;
    }
    newObj.canRead = obj.canRead;
    newObj.text = getTextObj(obj);
    newObj.shape = getShapeObj(obj);
    newObj.expanded = false;
    newObj.direction = dir;
    newObj.children = [];
    newObj.childrenCount = dir === -1 ? obj.superCount : obj.childCount;
    newArr.push(newObj);
  }
  return newArr;
};
/**
 * 初始化图形数据
 * @param obj okr对象
 */
const getShapeObj = (obj) => {
  const info = getScheduleInfo(obj);
  let arr = [
    {
      type: "circle",
      background: "rgba(27,196,137,0.10)",
      left: 30,
      top: 30,
      r: 15,
    },
  ];
  if (obj.canRead) {
    let rect = [
      {
        type: "rect",
        background: info[1],
        left: 0,
        top: 0,
        width: 3,
        height: 100,
      },
    ];
    arr = arr.concat(rect);
  }
  return arr;
};
/**
 * 初始化文字数据
 * @param obj okr对象
 */
const getTextObj = (obj) => {
  const info = getScheduleInfo(obj);
  let arr = [
    {
      value: obj.employeeName.slice(-1) || "",
      top: 24,
      left: 30,
      width: 14,
      height: 14,
      fontsize: 12,
      lineHeight: 12,
      textAlign: "center",
      color: "#1BC489",
    },
    {
      value: obj.employeeName || "",
      top: 20,
      left: 54,
      width: 60,
      height: 16,
      fontsize: 16,
      lineHeight: 16,
      color: "#666666",
    },
    {
      value: obj.departmentName || "",
      top: 20,
      left: 120,
      width: 70,
      height: 16,
      fontsize: 16,
      lineHeight: 16,
      color: "#666666",
    },
  ];
  if (obj.canRead) {
    let desc = [
      {
        value: info[0],
        top: 22,
        left: 310,
        width: 110,
        height: 14,
        fontsize: 14,
        lineHeight: 14,
        textAlign: "right",
        color: info[1],
        isShowAll:true,
      },
      {
        value: obj.objectiveDesc || "",
        top: 50,
        left: 54,
        width: 250,
        height: 40,
        fontsize: 14,
        lineHeight: 20,
        color: "#333333",
      },
    ];
    arr = arr.concat(desc);
  } else {
    let desc = [
      {
        value: "暂无权限",
        top: 60,
        left: 130,
        width: 250,
        height: 40,
        fontsize: 16,
        lineHeight: 20,
        color: "#999",
      },
    ];
    arr = arr.concat(desc);
  }
  return arr;
};
const getMindData = (data) => {
  initRoot(data);
  return parseOkrJSON(data);
};

var nodeArray = getMindData(array);
console.log(nodeArray);
